source/assets/js/playground/console-utils.ts

```diff
@@ -36,115 +36,112 @@ export type ConsoleLog = ConsoleLogDebug | ConsoleLogWarning | ConsoleLogError;
-      data.lineNumber = span.start.line;
-      data.range = [
+      lineNumber = span.start.line;
+      range = [
         span.start.line + 1,
         span.start.column + 1,
         span.end.line + 1,
         span.end.column + 1,
       ];
     }
-    data.message = item.error?.toString() || '';
+    message = item.error?.toString() || '';
   } else if (['debug', 'warn'].includes(item.type)) {
-    data.message = item.message;
-    let lineNumber = item.options.span?.start?.line;
+    message = item.message;
+    let _lineNumber = item.options.span?.start?.line;
     if (item.options.span) {
       const span = item.options.span;
-      data.lineNumber = span.start.line;
-      data.range = [
+      lineNumber = span.start.line;
+      range = [
         span.start.line + 1,
         span.start.column + 1,
         span.end.line + 1,
         span.end.column + 1,
       ];
     }
 
-    if (typeof lineNumber === 'undefined') {
+    if (typeof _lineNumber === 'undefined') {
       const stack = 'stack' in item.options ? item.options.stack : '';
       const needleFromStackRegex = /^- (\d+):(\d+) /;
       const match = stack?.match(needleFromStackRegex);
       if (match?.[1]) {
         // Stack trace starts at 1, all others come from span, which starts at
         // 0, so adjust before formatting.
-        lineNumber = parseInt(match[1]) - 1;
+        _lineNumber = parseInt(match[1]) - 1;
       }
       if (match?.[2]) {
-        data.range = [
+        range = [
           parseInt(match[1]),
           parseInt(match[2]),
           parseInt(match[1]),
           parseInt(match[2]),
         ];
       }
     }
-    data.lineNumber = lineNumber;
+    lineNumber = _lineNumber;
 
     if (item.type === 'warn' && item.options.deprecationType?.id) {
-      data.safeLink = `https://sass-lang.com/d/${item.options.deprecationType.id}`;
+      safeLink = `https://sass-lang.com/d/${item.options.deprecationType.id}`;
     }
   }
-  const link = selectionLink(playgroundState, data.range);
+  const link = selectionLink(playgroundState, range);
 
   const locationStart = link
-    ? `<a href="#${link}" class="console-location" data-range=${data.range}>`
+    ? `<a href="#${link}" class="console-location" data-range=${range}>`
     : '<div class="console-location">';
 
   const locationEnd = link ? '</a>' : '</div>';
 
+  let html = encodeHTML(message);
+
+  if (safeLink) {
+    // Wrap text link in anchor link
+    html = html.replace(
+      safeLink,
+      `<a href="${safeLink}" target="_blank">${safeLink}</a>`
+    );
+  }
+
   return `<div class="console-line">${locationStart}<span class="console-type console-type-${
-    data.type
-  }">@${data.type}</span>:${lineNumberFormatter(
-    data.lineNumber
-  )}${locationEnd}<div class="console-message">${encodeHTML(data.message, data.safeLink)}</div></div>`;
+    type
+  }">@${type}</span>:${lineNumberFormatter(
+    lineNumber
+  )}${locationEnd}<div class="console-message">${html}</div></div>`;
```

Comment:
Does this end up with a dangling `:` if `lineNumber` is undefined?

Reply:
It does, only if a warning does not have a span, or does not include a parseable line number in its stack. I wasn't sure how to replicate that to verify, but I fixed that in b738fd44297af0931d22eca1bb3981607097b4e0.